python/ql/src/experimental/Security/CWE-770/UnicodeDoS.ql

```diff
@@ -0,0 +1,121 @@
+/**
+ * @name Denial of Service using Unicode Characters
+ * @description A remote user-controlled data can reach a costly Unicode normalization with either form NFKC or NFKD. On Windows OS, with an attack such as the One Million Unicode Characters, this could lead to a denial of service. And, with the use of special Unicode characters, like U+2100 (℀) or U+2105 (℅), the payload size could be tripled.
+ * @kind path-problem
+ * @id py/unicode-dos
+ * @precision high
+ * @problem.severity error
+ * @tags security
+ *       experimental
+ *       external/cwe/cwe-770
+ */
+
+import python
+import semmle.python.ApiGraphs
+import semmle.python.Concepts
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.DataFlowPublic
+import semmle.python.dataflow.new.RemoteFlowSources
+
+// The Unicode compatibility normalization calls from unicodedata, unidecode, pyunormalize
+// and textnorm modules. The use of argIdx is to constraint the argument being normalized.
+class UnicodeCompatibilityNormalize extends API::CallNode {
+  int argIdx;
+
+  UnicodeCompatibilityNormalize() {
+    exists(API::CallNode cn, DataFlow::Node form |
+      cn = API::moduleImport("unicodedata").getMember("normalize").getACall() and
+      form.asExpr().(StrConst).getS() in ["NFKC", "NFKD"] and
+      TaintTracking::localTaint(form, cn.getArg(0)) and
+      this = cn and
+      argIdx = 1
+    )
```

Comment:
This can be written without the existential quantifications:
```suggestion
    this = API::moduleImport("unicodedata").getMember("normalize").getACall() and
    this.getParameter(0).getAValueReachingSink().asExpr().(StrConst).getText() in ["NFKC", "NFKD"] and
    argIdx = 1
```
we should probably do that for all the cases. You could then consider grouping the cases by how the parameters are supplied.

Reply:
Used the getAValueReachingSink() while avoiding the use of new vars and grouped the Calls by `argIdx` value https://github.com/github/codeql/pull/15319/commits/ebe599b6285b8634609f8353adbe95f214fd2e07